workers/unity/Packages/io.improbable.worker.sdk/IOStream.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.IO;
+using Improbable.Worker.CInterop.Internal;
+
+namespace Improbable.Worker.CInterop
+{
+    public enum OpenMode
+    {
+        /* Opens the stream in the default mode. */
+        OpenModeDefault = 0x00,
+    }
+
+    public sealed unsafe class IOStream : IDisposable
+    {
+        private readonly CIO.StreamHandle stream;
+
+        private IOStream(CIO.StreamHandle stream)
+        {
+            this.stream = stream;
+        }
+
+        /// <inheritdoc cref="IDisposable"/>
+        public void Dispose()
+        {
+            stream.Dispose();
+        }
+
+        public static IOStream CreateRingBufferStream(uint capacity)
+        {
+            return new IOStream(CIO.CreateRingBufferStream(capacity));
+        }
+
+        public static IOStream CreateFileStream(string fileName, OpenMode openMode)
+        {
+            fixed (byte* fileNameBytes = ApiInterop.ToUtf8Cstr(fileName))
+            {
+                return new IOStream(CIO.CreateFileStream(fileNameBytes, (CIO.OpenMode) openMode));
+            }
+        }
+
+        public long Write(byte[] data)
+        {
+            var remainingCapacity = CIO.StreamGetRemainingWriteCapacityBytes(stream);
+            if (remainingCapacity < data.Length)
+            {
+                throw new IOException("Not enough stream capacity to write data.");
```

Comment:
It may be worth have a different exception for this case as opposed to the general error case. It means users can write logic to specifically handle it.

It seems that [`NotSupportedException`](https://docs.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=netcore-3.1) may be a good candidate as its used elsewhere in the C# standard library for 'stream full' errors (e.g. - in [`Memory Stream`](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.write?view=netcore-3.1#System_IO_MemoryStream_Write_System_Byte___System_Int32_System_Int32_))